### DISCUSSION TITLE
Is re-frame for Clojure reactive programming? (answer: no, use nabab instead)

### DISCUSSION THREAD

**Discussion by piotr-yuxuan** [day8/re-frame#686] [Q&A]
Hello, thanks again for that great project :-)

I'm using it server-side to take part of [cljc isomorophism](https://en.wikipedia.org/wiki/Isomorphic_JavaScript) (run same code in server and client) and for the most part it's quite awesome but after reading some code and documentation, I wonder whether it's an intended usage.

> `re-frame` is a pattern for writing SPAs in ClojureScript, using Reagent.
> — From the README

> …there isn't often much point firing a timed event in a test.
> — From [Clojure code](https://github.com/Day8/re-frame/blob/64f858c4e33f68b1da5d6f9532defd8a73345b0b/src/re_frame/interop.clj#L78-L79)

From [Clojure code](https://github.com/Day8/re-frame/blob/64f858c4e33f68b1da5d6f9532defd8a73345b0b/src/re_frame/interop.clj#L24):
``` clj
(defonce ^:private executor (Executors/newSingleThreadExecutor))
```

Hence, I have a couple of questions:

- Why a single thread?
- Would you accept a PR to use multiple threads in the JVM?
- If any strong reasons not to use it server-side, would you know of any other libraries?

**Comment by danielcompton** [comment]
Hi @piotr-yuxuan, we added CLJ support for re-frame to support testing and development use-cases. This lets you simulate re-frame handlers calling to in-process 'server-side' handlers, run unit tests from a Clojure REPL, and similar dev-time activities. I've also seen re-frame used server-side to render an HTML page which is then hydrated with Reagent when it is sent over to the client.

I guess you could use re-frame for creating a server, but I'm not really sure how it would work. In re-frame the results of your events don't directly create a response, they only update a reactive (in ClojureScript) database which can then trigger subscriptions. You would need some way of replying to a client with the changes from an event. It doesn't seem like a terribly good fit to me, but I'm interested in hearing more about how you would do it.

To your questions:

> * Why a single thread?

Because re-frame was built to work in a single-threaded browser environment, and we wanted the Clojure version to match that.

> * Would you accept a PR to use multiple threads in the JVM?

Probably not, I don't think it's a path we're very interested in, and we've already had some concurrency issues on the JVM which we've had to fix.

> * If any strong reasons not to use it server-side, would you know of any other libraries?

I'd recommend looking at [CQRS](https://martinfowler.com/bliki/CQRS.html) and event sourcing patterns. There is [CMDR](https://github.com/capitalone/cqrs-manager-for-distributed-reactive-services) which might give you some ideas.

In summary, building server applications isn't re-frame's intended usage (currently), and you might find other solutions which are a better fit.

**Comment by piotr-yuxuan** [comment]
@danielcompton,

Thanks a lot for your detailed, precisely worded reply! Now my previous message looks blur, and I like to give it more context.

I've started [a pet project](https://github.com/piotr-yuxuan/choose-a-new-phone) which displays a list of LineageOS-compatible phones. I wrote it with re-frame. At first the phone list was constructed on each page rendering but page load was slow so UX wasn't great. I chose to fetch and insert the list at compile-time so it would faster the first rendering and still update it client-side to make sure it's up to date. In order to do so and minimize logic duplication, I ported my existing cljs re-frame code to cljc. Now I'm figuring out that some other data sources can't be fetched from a client runtime (because of CORS), only from a Clojure runtime (e.g a REPL) so this pet project is going to be executed more and more from a Clojure REPL.

Right now I'm standing here: I like the structured event-based programming style but I'm wondering whether re-frame is still the right tool for this job, hence the title of this issue. In my humble point of view, here are the pros and cons, in no particular order:

- 👍 re-frame avoids code duplication between clj and cljs, but duplication is better than bad abstraction
- 👍 re-frame code is about how we mutate data, has structured event-based programming style with strong conceptual model, and strong state management.
- ❌ re-frame is single-threaded both in cljs (of course) and in clj (by choice)
- ❌ clj and cljs fx differ, I'm maintaining a common API but is it a good thing?
- ❌ re-frame in clj sometimes has different behaviour (like `set-timeout!`)

If re-frame isn't a good fit for that use case, and if multi-threaded re-frame isn't interesting, would anybody mind suggest some structured event-based programming libraries? [Manifold](https://github.com/ztellman/manifold), [RxClojure](https://github.com/ReactiveX/RxClojure), or [quasar-pulsar](http://blog.paralleluniverse.co/2013/05/02/quasar-pulsar/)? I've had a brief look at them and they all seem to miss the "data-oriented, state management" part.

**Comment by mike-thompson-day8** [comment]
Further to what @danielcompton has written, I'd  recommend recomend two FAQs to you:
   - https://github.com/Day8/re-frame/blob/master/docs/FAQs/Why-CLJC.md
   - https://github.com/Day8/re-frame/blob/master/docs/FAQs/FullStackReframe.md

We'd like re-frame to be excellent for browser-based SPAs.  Beyond that, we have no ambitions.

**Comment by piotr-yuxuan** [comment]
@danielcompton, thank you for your reply.

**Comment by danielcompton** [comment]
If you'd like to get a list of phones pre-loaded into the page, you could look at bundling the JSON via a global script tag on the HTML page you serve to the browser, or maybe as a URL encoded JSON string inside a meta tag (I've seen this technique on some of Apple's sites).

Will close this for now, but thanks for the good question.

**Comment by piotr-yuxuan** [comment]
Follow-up: _use the right tool for the right job_. `re-frame` is the right tool for graphical FRP. I've switched to `core.async` + minimal plumbing for computation-oriented FRP. Thanks all of you for your precise replies 🙂

https://github.com/piotr-yuxuan/nabab
